View Asays, Erase Flavor Text and Use Random Name

dearmochi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Adds **Erase Flavor Text** action in the Player Panel to erase a player's flavor text, both from their body and their saved character in the database
Adds **Use Random Name** action in the Player Panel to force a player to use a random name - it will rename their body and datacore completely as well as save their new name in their character preferences
Adds **View Asays** verb in the Admin tab that displays all Asays in the current round

## Why It's Good For The Game
The new actions would help staff avoid having to warn players for the same flavor texts/names every time
It can be easy to miss Asays so adding a log would help in catching up

## Images of changes
![image](https://user-images.githubusercontent.com/1496804/83945909-f5cb3d00-a80d-11ea-9f7c-ad5e530d8644.png)

## Changelog
:cl:
add: Add ability for admins to erase a player's flavor text
add: Add ability for admins to force a player to use random name
add: Add Asay log for admins
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AffectedArc07 commented:
Few things. For the name randomisation, why not just have a button to randomise it in round, then also update and save their prefs to avoid it not being changed in the round. And for the asay logs, make the window wider, and **please** put some table borders.

dearmochi commented:
> 
> 
> Few things. For the name randomisation, why not just have a button to randomise it in round, then also update and save their prefs to avoid it not being changed in the round. And for the asay logs, make the window wider, and **please** put some table borders.

I thought about having the name randomisation button rename the body, but I'm not sure how to approach it with ID cards, PDAs and stuff. Should I just rename them manually as well?

AffectedArc07 commented:
Theres a proc on the mob, `/mob/proc/rename_character(oldname, newname)` which will auto rename stuff and update the datacore

dearmochi commented:
> 
> 
> Theres a proc on the mob, `/mob/proc/rename_character(oldname, newname)` which will auto rename stuff and update the datacore (Manifest and stuff, not to be confused with database)

Handy, thanks! I've added that in and removed the part where it sets the character preferences to always use a random name - instead, it will save that random name. I also added borders to the table

Fox-McCloud commented on code/modules/admin/topic.dm:
```diff
@@ -1287,6 +1287,57 @@
+		message_admins("[key_name_admin(usr)] has erased [key_name_admin(M)]'s flavor text.")
+
+		// Clears the mob's flavor text
+		M.flavor_text = "";
```
what's up with the erroneous `;`

dearmochi replied:
The result of knowing more languages than one person should in their lifetime

Fox-McCloud commented on code/modules/admin/topic.dm:
```diff
@@ -1287,6 +1287,57 @@
+		M.flavor_text = "";
+
+		// Clear and save the DB character's flavor text
+		M.client.prefs.flavor_text = "";
```
again

AffectedArc07 replied:
DM accepts `;`?

dearmochi replied:
removing

farie82 commented on code/modules/admin/verbs/asays.dm:
```diff
@@ -0,0 +1,63 @@
+	var/time = 0
+
+/client/proc/view_asays()
+	set name = "View Asays"
```
I'd change this to "Asays" only or such. Admins will look for asays and thus look at the A first. Atleast that's what I heard from some of them.

dearmochi replied:
Good point

farie82 replied:
Still applies

dearmochi replied:
Whoops, fixed

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1287,6 +1287,57 @@
+
+		var/mob/M = locateUID(href_list["eraseflavortext"])
+
+		if(!M.client)
```
Best to also add a null check on `M`. It is an edge case but it can happen.

dearmochi replied:
Alright, I was copying the pattern as the other topics which don't have a null check, so I'll add it on the other ones as well

farie82 replied:
Might have forgotten to push the changes. Else it's best to keep comments unresolved till they are actually resolved on the PR itself

dearmochi replied:
Aye, juggling multiple PRs at the same time is a recipe for a disaster, I'll get to it now. Thanks

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1287,6 +1287,57 @@
+
+		var/mob/M = locateUID(href_list["userandomname"])
+
+		if(!M.client)
```
Same here

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1287,14 +1288,73 @@
+		var/mob/M = locateUID(href_list["userandomname"])
+
+		if(!istype(M, /mob))
+			to_chat(usr, "This can only be used on instances of type /mob")
```
Best to use a span here

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1287,14 +1288,73 @@
+		var/mob/M = locateUID(href_list["eraseflavortext"])
+
+		if(!istype(M, /mob))
+			to_chat(usr, "This can only be used on instances of type /mob")
```
Best to use a span here

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1548,6 +1608,11 @@
 		var/mob/M = locateUID(href_list["adminplayeropts"])
+
+		if(!istype(M, /mob))
+			to_chat(usr, "This can only be used on instances of type /mob")
```
same

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1557,6 +1622,11 @@
 		var/mob/M = locateUID(href_list["adminplayerobservefollow"])
+
+		if(!istype(M, /mob))
+			to_chat(usr, "This can only be used on instances of type /mob")
```
same

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1630,6 +1700,11 @@
 		var/mob/M = locateUID(href_list["adminmoreinfo"])
+
+		if(!istype(M, /mob))
+			to_chat(usr, "This can only be used on instances of type /mob")
```
same. Also goes for the others below here

farie82 commented on code/modules/admin/verbs/asays.dm:
```diff
@@ -0,0 +1,63 @@
+GLOBAL_LIST_EMPTY(asays)
+
+/datum/asays
+	var/ckey = ""
```
Best to not init them here since you will init them all manually later. Making a `New` for this will be nice as well so you don't have to manually set all the vars after newing it

dearmochi replied:
Good point, I've fixed it just now

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -747,7 +747,7 @@
 
 		if(!SSjobs)
-			to_chat(usr, "SSjobs has not been setup!")
+			to_chat("<span class='warning'>SSjobs has not been setup!</span>")
```
missing `usr` here

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -488,14 +488,14 @@
 			return
 		if(!SSjobs)
-			to_chat(usr, "SSjobs has not been setup!")
+			to_chat("<span class='warning'>SSjobs has not been setup!</span>")
```
missing `usr` here

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -437,17 +437,17 @@
 		if(banreason)
 	/*		if(!config.ban_legacy_system)
-				to_chat(usr, "Unfortunately, database based unbanning cannot be done through this panel")
+				to_chat("<span class='warning'>Unfortunately, database based unbanning cannot be done through this panel</span>")
```
missing `usr` here. It's a comment though. Can also just be removed

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -114,39 +114,39 @@
 			if(BANTYPE_PERMA)
 				if(!banckey || !banreason)
-					to_chat(usr, "Not enough parameters (Requires ckey and reason)")
+					to_chat("<span class='warning'>Not enough parameters (Requires ckey and reason)</span>")
```
missing all the `usr` vars here. Best to go over the commit again and re add the usr params

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1491,7 +1491,7 @@
 		var/mob/living/L = locateUID(href_list["revive"])
 		if(!istype(L))
-			to_chat(usr, "This can only be used on instances of type /mob/living")
+			to_chat(usr, "<span class='warning'>This can only be used on instances of type /mob</span>/living")
```
No `</span>`s here and below. Best to recheck the commit

farie82 commented on code/modules/admin/topic.dm:
```diff
@@ -1712,7 +1713,7 @@
 		var/mob/living/carbon/human/H = locateUID(href_list["adminspawncookie"])
 		if(!ishuman(H))
-			to_chat(usr, "This can only be used on instances of type /mob/living/carbon/human")
+			to_chat(usr, "<span class='warning'>This can only be used on instances of type /mob</span>/living/carbon/human")
```
Spans incorrectly added here

dearmochi replied:
Whoops, I'll be more careful when mass-replacing next time. All done

TDSSS commented:
Does this window refresh if a new asay comes in? If not, a refresh button might be nice, like the PMs window has.

AffectedArc07 commented on code/modules/admin/verbs/asays.dm:
```diff
@@ -0,0 +1,69 @@
+	if(!check_rights(R_ADMIN))
+		return
+
+	var/output = {"
+		<style>
+		td, th
+		{
+			border: 1px solid #425c6e;
+			padding: 3px;
+		}
+
+		thead
+		{
+			color: #517087;
+			font-weight: bold;
+			table-layout: fixed;
+		}
+	</style>
+	<table style='width: 100%; border-collapse: collapse; table-layout: auto;'>
+	"}
+
+	// Header & body start
+	output += {"
+		<thead>
+			<tr>
+				<th>Time</th>
+				<th>Ckey</th>
+				<th>Message</th>
+			</tr>
+		</thead>
+		<tbody>
+	"}
```
Any reason these are separate vars?

AffectedArc07 commented on code/modules/admin/verbs/asays.dm:
```diff
@@ -0,0 +1,69 @@
+
+	for (var/datum/asays/A in GLOB.asays)
+		var/timestr = time2text(A.time, "hh:mm:ss")
+		output += {"
+			<tr>
+				<td>[timestr]</td>
+				<td><b>[A.ckey] ([A.rank])</b></td>
+				<td>[A.message]</td>
+			</tr>
+		"}
```
String concatenation is quite expensive in BYOND. Make `output` a list and then do this further down

```
	var/html_string = output.Join()
	popup.set_content(html_string)
```

AffectedArc07 commented:
Tested this locally, the two resets work, but I have a request for the asay panel

1: A refresh button
2: Better table widths. I would use 5% for time, 10% for ckey, and 85% for the message column.

dearmochi commented:
All done

![image](https://user-images.githubusercontent.com/1496804/89760296-62073d00-daec-11ea-897c-7ed3fb1f2c39.png)

AffectedArc07 commented on code/modules/admin/topic.dm:
```diff
@@ -1413,11 +1479,11 @@
 		var/mob/M = locateUID(href_list["aroomwarp"])
-		if(!ismob(M))
-			to_chat(usr, "This can only be used on instances of type /mob")
+		if(!istype(M, /mob))
```
Any reason you overrode all the instances of `ismob`?

dearmochi replied:
I think this was done when I was newer to DM and thought `ismob` required a non-null argument, but that's obviously not the case after testing.
